### PR TITLE
Make LineUp v4 RegExp filter serialization backwards compatible

### PR DESCRIPTION
**Pull request by oltionchampari**
Close datavisyn/tdp_core#331

## Summary 

* Extended tests for `serializeLineUpFilter()` and `restoreRegExp()` functions.
* Refactored the functions to be backwards compatible.

### CODE REVIEW

File: tests/lineup/internal/cmds.test.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  });
+
+  it('filter as null', () => {
+    const lineUpFilter = {filter: null, filterMissing: false};
+    expect(serializeLineUpFilter(lineUpFilter)).toMatchObject({filter: {value: null, isRegExp: false}, filterMissing: false});
+  });
+
+  it('filter as null and use regular expression true', () => {
+    const lineUpFilter = {filter: /null/, filterMissing: false};
+    expect(serializeLineUpFilter(lineUpFilter)).toMatchObject({filter: {value: /null/, isRegExp: true}, filterMissing: false});
```
Comment (oltionchampari) [inline tests/lineup/internal/cmds.test.ts]: What should happen in this case ?
serialize

File: src/lineup/internal/cmds.ts [hunk 2]
```diff
@@ -378,49 +389,98 @@ interface IRegExpFilter {
...
  */
-function restoreRegExp(filter: string | IRegExpFilter): string | RegExp {
-  if (filter === null || !(<IRegExpFilter>filter).isRegExp) {
-    return <string | null>filter;
+export function restoreRegExp(filter: ILineUpStringFilterValue | IRegExpFilter | ISerializableLineUpFilter, filterMissing = false): ILineUpStringFilter {
+  const isIRegExpFilter = ((filter: IRegExpFilter | ISerializableLineUpFilter): filter is IRegExpFilter => filter.hasOwnProperty('isRegExp'));
+  const isISerializableLineUpFilter = (filter: IRegExpFilter | ISerializableLineUpFilter): filter is ISerializableLineUpFilter => filter.hasOwnProperty('filterMissing');
+  if (filter === null || typeof filter === 'string' || Array.isArray(filter)) {
+    return {filter: <string | null>filter, filterMissing};
+  } else if (isIRegExpFilter(filter)) {
```
Comment (thinkh) [inline src/lineup/internal/cmds.ts]: Just a minor code style. I would add a new line before each else block

```suggestion

  } else if (isIRegExpFilter(filter)) {
```

File: tests/lineup/internal/cmds.test.ts (hunk 1 above)
Comment (oltionchampari) [reply tests/lineup/internal/cmds.test.ts]: After a second inspection i found out that the value  `/null/` is not possible since the change regex filter checkbox reacts only if a value is present.

I added another test for the case when you click the filter missing values and filter is empty.

filter

File: src/lineup/internal/cmds/filter.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+
+/**
+ * This type guard checks if a given parameter has the `filter` property.
+ * Necessary since number columns filter has properties `min`, `max` and no filter property,
+ * @interal
+ *
+ * @param filter Any value that could be a filter
+ */
+export function isSerializedFilter(filter: any): ISerializableLineUpFilter {
```
Comment (thinkh) [inline src/lineup/internal/cmds/filter.ts:62]:
```suggestion
export function isSerializedFilter(filter: any): filter is ISerializableLineUpFilter {
```

I would expect that this is a type guard function. And the `return` statement indicates a boolean return value. Please revise.

File: src/lineup/internal/cmds/filter.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ */
+type ILineUpStringFilterValue = string[] | string | null;
+
+
+/**
+ * This type guard checks if a given parameter has the `filter` property.
+ * Necessary since number columns filter has properties `min`, `max` and no filter property,
+ * @interal
+ *
+ * @param filter Any value that could be a filter
```
Comment (thinkh) [inline src/lineup/internal/cmds/filter.ts:60]: `@returns` documentation is missing.

File: tests/lineup/internal/cmds.test.ts (hunk 1 above)
Comment (thinkh) [reply tests/lineup/internal/cmds.test.ts]: Ok, good. Then we do not need to unit test this.

File: src/lineup/internal/cmds/filter.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    return restoreLineUpFilter(filter.value, filterMissing);
+
+  } else if (isISerializableLineUpFilter(filter)) {
+    return restoreLineUpFilter(filter.filter, filter.filterMissing);
+
+  }
+
+  throw new Error('Unknown LineUp filter format. Unable to restore the given filter.');
```
Comment (thinkh) [inline src/lineup/internal/cmds/filter.ts:125]: I added this error message to prevent cases where the return value of this function is `void`. After playing around with this implementation I got an error with the regexp string:

ordino-lineup-unknown-filter-format

This is the error in the console:

```
filter.ts:125 Uncaught (in promise) Error: Unknown LineUp filter format. Unable to restore the given filter.
    at restoreLineUpFilter (filter.ts:125)
    at restoreLineUpFilter (filter.ts:121)
    at ActionNode.<anonymous> (cmds.ts:227)
    at Generator.next (<anonymous>)
    at fulfilled (tslib.es6.js:71)
```

@oltionchampari Can you check why calling it recursively (starting from line 121) and not returning `void` (without the error) and now we run into the error? In guess is that we still have an uncovered case where the three typeguards above are not working correctly.

My expectation would be that we do not get the error message and one of the typeguards catches the recursive call and returns a valid filter object.

Comment (oltionchampari) [reply src/lineup/internal/cmds/filter.ts:125]: The issue was that when you go back in the provenance graph the `restoreLineUpFilter()`
function gets called with the old value which gets derived from the Ranking.Thus it hasn't been serialized:

 ```typescript
const fromGraphFilter = {filter: /abc/, filterMissing: false};
   restoreLineUpFilter(fromGraphFilter)
```
I added a test to cover this case and returned the filter when it's already serialized.

File: src/lineup/internal/cmds/filter.ts (hunk 3 above)
Comment (oltionchampari) [reply src/lineup/internal/cmds/filter.ts]: The function checks if the filter is `ISerializableLineUpFilter | ILineUpStringFilter`, so i added a typeguard that includes both.

Comment (thinkh) [reply src/lineup/internal/cmds/filter.ts]: That fix was only part of the solution. The reason why you had to add `ILineUpStringFilter` to everything was that we were using the function `isSerializedFilter` in `recordPropertyChange` which is incorrect! Instead we must check if the provided filter is a valid `ILineUpStringFilter`. I refactored the code and extracted + tested the type guards.

**Comment by thinkh** [comment]
@oltionchampari I refactored the code again and wrote more type guards and tests. For me it works, but can you please check and test the solution in the browser again. If you do not find any errors or flaws we are ready to merge.

**Comment by oltionchampari** [comment]
It works as expected for me too, i see no problems with it. The code is much better organized now.
